src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+        SourceGeneratedDocumentIdentity documentIdentity, SourceText sourceText)
+    {
+        // We won't support freezing multiple source generated documents at once. Although nothing in the implementation
+        // of this method would have problems, this simplifies the handling of serializing this solution to out-of-proc.
+        // Since we only produce these snapshots from an open document, there should be no way to observe this, so this assertion
+        // also serves as a good check on the system. If down the road we need to support this, we can remove this check and
+        // update the out-of-process serialization logic accordingly.
+        Contract.ThrowIfTrue(_frozenSourceGeneratedDocumentState != null, "We shouldn't be calling WithFrozenSourceGeneratedDocument on a solution with a frozen source generated document.");
+
+        var existingGeneratedState = TryGetSourceGeneratedDocumentStateForAlreadyGeneratedId(documentIdentity.DocumentId);
+        SourceGeneratedDocumentState newGeneratedState;
+
+        if (existingGeneratedState != null)
+        {
+            newGeneratedState = existingGeneratedState
+                .WithText(sourceText)
+                .WithParseOptions(existingGeneratedState.ParseOptions);
+
+            // If the content already matched, we can just reuse the existing state
+            if (newGeneratedState == existingGeneratedState)
+            {
+                return this;
+            }
+        }
+        else
+        {
+            var projectState = this.Solution.GetRequiredProjectState(documentIdentity.DocumentId.ProjectId);
+            newGeneratedState = SourceGeneratedDocumentState.Create(
+                documentIdentity,
+                sourceText,
+                projectState.ParseOptions!,
+                projectState.LanguageServices,
+                // Just compute the checksum from the source text passed in.
+                originalSourceTextChecksum: null);
+        }
+
+        var projectId = documentIdentity.DocumentId.ProjectId;
+        var newTrackerMap = CreateCompilationTrackerMap(projectId, this.Solution.GetProjectDependencyGraph());
+
+        // We want to create a new snapshot with a new compilation tracker that will do this replacement.
+        // If we already have an existing tracker we'll just wrap that (so we also are reusing any underlying
+        // computations). If we don't have one, we'll create one and then wrap it.
+        if (!newTrackerMap.TryGetValue(projectId, out var existingTracker))
+        {
+            existingTracker = CreateCompilationTracker(projectId, this.Solution);
+        }
+
+        newTrackerMap = newTrackerMap.SetItem(
+            projectId,
+            new GeneratedFileReplacingCompilationTracker(existingTracker, newGeneratedState));
+
+        return this.Branch(
+            // TODO(cyrusn): Is it ok to just pass this.Solution along here?
+            this.Solution,
```

Comment:
similar interesting case.

Reply:
Yep, same thing. As we go to updating these as generation runs asynchronously, we'll see those new "updates" landing in the same way as these: we're updating the source generated documents and compilations, but the underlying base state is the same.

Reply:
I'd also imagine this implementation changes drastically, since "here's just a set of documents that are frozen and are synced back to OOP" is a common thing rather than a special case.